backend/src/api/routers/chains.py

```diff
@@ -0,0 +1,169 @@
+    embeddings_model_name="BAAI/bge-large-en-v1.5",
+    reranking_model_name="BAAI/bge-reranker-base",
+    contextual_rerank=True,
+    use_cuda=True,
+    docs_path=["./data/markdown/ORFS_docs", "./data/markdown/OR_docs"],
+    manpages_path=["./data/markdown/manpages"],
+)
+hybrid_retriever.create_hybrid_retriever()
+hybrid_llm_chain = hybrid_retriever.get_llm_chain()
+
+sim_retriever_chain = SimilarityRetrieverChain(
+    llm_model=llm,
+    prompt_template_str=prompt_template_str,
+    embeddings_model_name="BAAI/bge-large-en-v1.5",
+    use_cuda=True,
+    docs_path=["./data/markdown/ORFS_docs", "./data/markdown/OR_docs"],
+    manpages_path=["./data/markdown/manpages"],
+)
+sim_retriever_chain.create_similarity_retriever()
+sim_llm_chain = sim_retriever_chain.get_llm_chain()
+
+multi_retriever_chain = MultiRetrieverChain(
+    llm_model=llm,
+    prompt_template_str=prompt_template_str,
+    embeddings_model_name="BAAI/bge-large-en-v1.5",
+    use_cuda=True,
+    docs_path=["./data/markdown/ORFS_docs", "./data/markdown/OR_docs"],
+    manpages_path=["./data/markdown/manpages"],
+)
+multi_retriever_chain.create_multi_retriever()
+multi_llm_chain = multi_retriever_chain.get_llm_chain()
+
+@router.get("/listAll")
+async def list_all_chains() -> list:
+    return ["/chains/hybrid","/chains/sim","/chains/ensemble"]
+
+@router.post("/hybrid")
+async def get_hybrid_response(user_input: UserInput) -> dict:
+    user_question = user_input.query
+    result = hybrid_llm_chain.invoke(user_question)
+
+    links = []
+    context = []
+    for i in result["context"]:
+        if "url" in i.metadata:
+            links.append(i.metadata["url"])
+        elif "source" in i.metadata:
+            links.append(i.metadata["source"])
+        context.append(i.page_content)
+
+    links_set = set(links)
+
+    if user_input.list_sources and user_input.list_context:
+        response = {
+            "response": result["answer"],
+            "sources": (links_set),
+            "context": (context),
+        }
+    elif user_input.list_sources:
+        response = {"response": result["answer"], "sources": (links_set)}
+    elif user_input.list_context:
+        response = {"response": result["answer"], "context": (context)}
+    else:
+        response = {"response": result["answer"]}
+
+    return response
+
+@router.post("/sim")
+async def get_sim_response(user_input: UserInput) -> dict:
+    user_question = user_input.query
+    result = hybrid_llm_chain.invoke(user_question)
+
+    links = []
+    context = []
+    for i in result["context"]:
+        if "url" in i.metadata:
+            links.append(i.metadata["url"])
+        elif "source" in i.metadata:
+            links.append(i.metadata["source"])
+        context.append(i.page_content)
+
+    links_set = set(links)
+
+    if user_input.list_sources and user_input.list_context:
+        response = {
+            "response": result["answer"],
+            "sources": (links_set),
```

Comment:
Can a `set` be passed as input to this json? Or must you convert to string

Reply:
Same question for `list`

Reply:
I think its implicitly getting converted into a list here. I've edited this part to manually type cast the set to a list now.